types/wlr_drm_lease_v1.c

```diff
@@ -0,0 +1,680 @@
+	assert(wl_resource_instance_of(resource,
+				&wp_drm_lease_v1_interface, &lease_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void drm_lease_v1_destroy(struct wlr_drm_lease_v1 *lease) {
+	if (!lease) {
+		return;
+	}
+
+	wlr_log(WLR_DEBUG, "Destroying lease %"PRIu32, lease->lessee_id);
+
+	if (lease->lessee_id != 0) {
+		wlr_drm_lease_v1_revoke(lease);
+		return;
+	}
+
+	wl_list_remove(&lease->link);
+	wl_resource_set_user_data(lease->resource, NULL);
+
+	free(lease->connectors);
+	free(lease);
+}
+
+static void drm_lease_request_v1_destroy(
+		struct wlr_drm_lease_request_v1 *request) {
+	if (!request) {
+		return;
+	}
+
+	wlr_log(WLR_DEBUG, "Destroying request %p", request);
+
+	wl_list_remove(&request->link);
+	wl_resource_set_user_data(request->resource, NULL);
+
+	free(request->connectors);
+	free(request);
+}
+
+static void drm_lease_connector_v1_destroy(
+		struct wlr_drm_lease_connector_v1 *connector) {
+	if (!connector) {
+		return;
+	}
+
+	wlr_log(WLR_DEBUG, "Destroying connector %s", connector->output->name);
+
+	struct wl_resource *resource, *tmp;
+	wl_resource_for_each_safe(resource, tmp, &connector->resources) {
+		wp_drm_lease_connector_v1_send_withdrawn(resource);
+		wl_resource_set_user_data(resource, NULL);
+		wl_list_remove(wl_resource_get_link(resource));
+		wl_list_init(wl_resource_get_link(resource));
+	}
+
+	wl_list_remove(&connector->link);
+	wl_list_remove(&connector->destroy.link);
+	free(connector);
+}
+
+static void drm_lease_device_v1_destroy(
+		struct wlr_drm_lease_device_v1 *device) {
+	if (!device) {
+		return;
+	}
+
+	struct wlr_drm_backend *backend =
+			get_drm_backend_from_backend(device->backend);
+	wlr_log(WLR_DEBUG, "Destroying wlr_drm_lease_device_v1 for %s",
+			backend->name);
+
+	struct wl_resource *resource, *tmp_resource;
+	wl_resource_for_each_safe(resource, tmp_resource, &device->resources) {
+		wl_list_remove(wl_resource_get_link(resource));
+		wl_list_init(wl_resource_get_link(resource));
+		wl_resource_set_user_data(resource, NULL);
+	}
+
+	struct wlr_drm_lease_request_v1 *request, *tmp_request;
+	wl_list_for_each_safe(request, tmp_request, &device->requests, link) {
+		drm_lease_request_v1_destroy(request);
+	}
+
+	struct wlr_drm_lease_v1 *lease, *tmp_lease;
+	wl_list_for_each_safe(lease, tmp_lease, &device->leases, link) {
+		wlr_drm_lease_v1_revoke(lease);
+	}
+
+	struct wlr_drm_lease_connector_v1 *connector, *tmp_connector;
+	wl_list_for_each_safe(connector, tmp_connector, &device->connectors, link) {
+		drm_lease_connector_v1_destroy(connector);
+	}
+
+	wl_list_remove(&device->link);
+	wlr_global_destroy_safe(device->global, device->manager->display);
+
+	free(device);
+}
+
+struct wlr_drm_lease_v1 *wlr_drm_lease_request_v1_grant(
+		struct wlr_drm_lease_request_v1 *request) {
+	assert(request->lease);
+
+	wlr_log(WLR_DEBUG, "Attempting to grant request %p", request);
+
+	struct wlr_drm_lease_v1 *lease = request->lease;
+	if (request->invalid) {
```

Comment:
This should never happen, `drm_lease_request_v1_handle_submit` guards against this I think? We could replace this with an `assert`.